src/DbPostgreSQLAdapter/Database.cpp

```diff
@@ -0,0 +1,134 @@
+		}
+
+		const auto& table = m_tables.at(tableName);
+		return *(table.get());
+	}
+
+	std::unique_ptr<IRecordSet> Database::executeQuery(const std::string& query)
+	{
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, query.data()));
+		if (PQresultStatus(statementResult.get()) == PGRES_TUPLES_OK)
+		{
+			return std::make_unique<RecordSet>(statementResult.get());
+		}
+		else
+		{
+			const std::string errorMessage = PQresultErrorMessage(statementResult.get());
+			
+			std::ostringstream exceptionStrem;
+			exceptionStrem << "# ERR: SQLException in " << __FILE__;
+			exceptionStrem << "(Database::executeQuery) on line " << __LINE__ << '\n';
+			exceptionStrem << "# ERR: " << errorMessage << std::endl;
+
+			throw std::runtime_error(exceptionStrem.str());
+		}
+	}
+
+	std::unique_ptr<ITableRecordSet> Database::executeTableQuery(const std::string& query, ITable& table)
+	{	
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, query.data()));
+		if (PQresultStatus(statementResult.get()) == PGRES_TUPLES_OK)
+		{
+			return std::make_unique<TableRecordSet>(table, statementResult.get());
+		}
+		else
+		{
+			const std::string errorMessage = PQresultErrorMessage(statementResult.get());
+
+			std::ostringstream exceptionStrem;
+			exceptionStrem << "# ERR: SQLException in " << __FILE__;
+			exceptionStrem << "(Database::executeOperation) on line " << __LINE__ << '\n';
+			exceptionStrem << "# ERR: " << errorMessage << std::endl;
+
+			throw std::runtime_error(exceptionStrem.str());
+		}
+
+	}
+
+	void Database::executeOperation(const std::string& operation)
+	{
+		std::lock_guard<std::recursive_mutex> lock(m_mutex);
+		const auto statementResult = createRAIIPGresult(PQexec(m_database, operation.data()));
+		const auto result = PQresultStatus(statementResult.get());
+		if (result == PGRES_TUPLES_OK)
+		{
+			// Retrieve last inserted id. Requires that the query returns the id
+			const unsigned int rowsCount = static_cast<unsigned int>(PQntuples(statementResult.get()));
+			const std::string rowId = PQgetvalue(statementResult.get(), 0, rowsCount-1);
+			m_lastInsertedRowId = std::stoi(rowId);
+		}
+		else if (result != PGRES_COMMAND_OK)
+		{
+			const std::string errorMessage = PQresultErrorMessage(statementResult.get());
+
+			std::ostringstream exceptionStrem;
+			exceptionStrem << "# ERR: SQLException in " << __FILE__;
+			exceptionStrem << "(Database::executeOperation) on line " << __LINE__ << '\n';
+			exceptionStrem << "# ERR: " << errorMessage << std::endl;
+
+			throw std::runtime_error(exceptionStrem.str());
+		}
+
+		m_lastOperationRowsAffected = std::atoi(PQcmdTuples(statementResult.get()));
+	}
+
+	void Database::executeMultipleStatements(const std::string& statements)
+	{
+		throw std::runtime_error("Not implemented yet");
```

Comment:
I wonder if you could rely on `static_assert` to get compilation error when using a non implemented method instead of relying on the `throw` which you would not notice until execution.
In theory would be something like:
```
void Database::executeMultipleStatements(const std::string& statements)
{
    static_assert(false, "Not implemented yet");
    // As it is void I don't think you need to put the throw. If it wasn't void we should try.
}
```

Reply:
This is a well meaning idea, but the way static_assert works is stopping the process during the compilation. There is a runtime equivalent called assert, but it works only on debug.